### PR TITLE
Update mpl CMake scripts

### PR DESCRIPTION
**Pull request by jacobmerson**
CMake has been udpated to the "modern" style including export targets for the mpl interface library. Tests have also been updated to run using CTest.

Currently there is no "version" of mpl, so I set the version in the CMake script to 0.0.1, but this can be updated.

as per #11 I'd like to make a spack package as well. These packages can work off the git branches, but everything works much better if actual releases are made so that people depending on mpl can lock down the exact version. I will make the package for spack once the updated CMake gets merged.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,8 +1,54 @@
...
+add_library(mpl INTERFACE)
+target_include_directories(mpl
+  INTERFACE
+  $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>
+  $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>
+  )
+
+# convention for allowing use as a subdirectory
+add_library(mpl::mpl ALIAS mpl)
+target_link_libraries(mpl INTERFACE Threads::Threads MPI::MPI_CXX)
```
[the remark this answers is not in the document]
Comment (jacobmerson) [reply CMakeLists.txt:29]: It's a bit confusing, but if I interpret the [CMake docs ](https://cmake.org/cmake/help/latest/module/FindMPI.html) correctly then the `lang` should be `CXX` which refers to using the C library from C++. So, `MPI::MPI_<lang>` becomes `MPI::MPI_CXX`, whereas linking to the depreciated C++ API would be `MPI::MPI_MPICXX`.

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,8 +1,54 @@
...
+
+option(MPL_BUILD_EXAMPLES ON "build the mpl examples")
```
[the remark this answers is not in the document]
Comment (jacobmerson) [reply CMakeLists.txt:31]: I agree, I was trying to leave the behavior the same as the previous CMake. I will update this to default to `OFF`

File: CMakeLists.txt [hunk 3]
```diff
@@ -1,8 +1,54 @@
...
+# 3.9 required for mpi targets, 3.5 required for boost targets
+cmake_minimum_required(VERSION 3.9) 
 
-project(mpl)
+# FIXME I didn't see versioning info on github,
+# but semantic versioning should be considered
+project(mpl VERSION 0.0.1 LANGUAGES CXX C)
 
-set(ALLOW_DUPLICATE_CUSTOM_TARGETS TRUE)
+include(CTest)
```
[the remark this answers is not in the document]
Comment (jacobmerson) [reply CMakeLists.txt:8]: I will update this.

File: CMakeLists.txt [hunk 4]
```diff
@@ -1,8 +1,54 @@
...
 
-add_subdirectory(examples)
-add_subdirectory(test)
+# project requires c++11 to build
+set(CMAKE_CXX_STANDARD 11)
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
+set(CMAKE_CXX_EXTENSIONS OFF)
+
+find_package(MPI REQUIRED)
+find_package(Threads REQUIRED)
```
[the remark this answers is not in the document]
Comment (jacobmerson) [reply CMakeLists.txt:18]: Yeah I think threads are used in the ISend/IRecv functions. I get link errors if I do not add the threads library in CMake.

For example when I use `VERBOSE=1 make` the link line for `stl_container` is as follows without threads. As you can se `pthread` is not on the link line.

```
[ 56%] Linking CXX executable stl_container
cd /lore/mersoj/mpl-testing/mpl/build/examples && /opt/scorec/cmake/3.20.0/bin/cmake -E cmake_link_script CMakeFiles/stl_container.dir/link.txt --verbose=1
/lore/mersoj/spack/spack/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/gcc-7.5.0-kqbvj3xkm57p2e7cdkhbfuyrvl4g7j2u/bin/g++ -Wl,-rpath -Wl,/lore/mersoj/spack/spack/opt/spack/linux-rhel7-x86_64/gcc-7.5.0/mpich-3.3.2-zhnl7s6eq7xtyjtzxkab2o5lgvrye7ny/lib -L/lore/mersoj/spack/spack/opt/spack/linux-rhel7-x86_64/gcc-7.5.0/hwloc-2.4.1-gi5lwvwumbpdaa662354yaesyx2xy3rs/lib CMakeFiles/stl_container.dir/stl_container.cc.o -o stl_container  -Wl,-rpath,/lore/mersoj/spack/spack/opt/spack/linux-rhel7-x86_64/gcc-7.5.0/mpich-3.3.2-zhnl7s6eq7xtyjtzxkab2o5lgvrye7ny/lib /lore/mersoj/spack/spack/opt/spack/linux-rhel7-x86_64/gcc-7.5.0/mpich-3.3.2-zhnl7s6eq7xtyjtzxkab2o5lgvrye7ny/lib/libmpicxx.so /lore/mersoj/spack/spack/opt/spack/linux-rhel7-x86_64/gcc-7.5.0/mpich-3.3.2-zhnl7s6eq7xtyjtzxkab2o5lgvrye7ny/lib/libmpi.so 
/lore/mersoj/spack/spack/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/binutils-2.36.1-4zmigg7hwzybf334wpcoaw2zewnkxcqq/bin/ld: CMakeFiles/stl_container.dir/stl_container.cc.o: undefined reference to symbol 'pthread_create@@GLIBC_2.2.5'
/lore/mersoj/spack/spack/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/binutils-2.36.1-4zmigg7hwzybf334wpcoaw2zewnkxcqq/bin/ld: /lib64/libpthread.so.0: error adding symbols: DSO missing from command line
```

**Comment by rabauke** [comment]
Many thanks.

**Pull request merged**